File: python/src/circle_ai/aethernet/adapters.py
# adapters.py
#
# Ports of the CircleAI <-> AetherNet bridge adapters (C# — the EXACT spec):
#   AetherNetContextAdapter.cs          -> AetherNetContextAdapter
#   AetherNetTelemetryAdapter.cs        -> AetherNetTelemetryAdapter (+ ObserverBridge)
#   AetherNetDirectiveSink.cs           -> AetherNetDirectiveSink
#   AetherNetInboundDirectiveBridge.cs  -> AetherNetInboundDirectiveBridge
#   CircleAiAetherNetAiProvider.cs      -> CircleAiAetherNetAiProvider
#
# Each bridges CircleAI.Aether contracts to the AetherNet runtime seam
# (circle_ai.aethernet.extensibility). The translation is the load-bearing
# behaviour; the runtime itself is injected.

from __future__ import annotations

from typing import List, Mapping, Optional, Sequence

from ..aether.context import (
    AetherInstallLevel,
    AetherVersion,
    IAetherContext,
    VersionLike,
)
from ..aether.intelligence import IAetherIntelligence
from ..aether.security_layer import (
    ISecurityDirectiveConsumer,
    SecurityDirective,
)
from ..aether.telemetry import (
    IAetherTelemetry,
    IAetherTelemetryObserver,
    IDisposable,
)
from .event_translator import EventTranslator
from .extensibility import (
    AiNetworkHealthReport,
    AiRouteSuggestion,
    AiThreatLevel,
    AetherNetNetworkEvent,
    AetherNetNodeEvent,
    AetherNetRouteEvent,
    AetherNetSecurityDirective,
    AetherNetSecurityEvent,
    AetherNetTransportEvent,
    CURRENT_PROTOCOL_VERSION,
    IAetherNetAiProvider,
    IAetherNetSecurityDirectiveConsumer,
    IAetherNetTelemetry,
    IAetherNetTelemetryObserver,
    MeshPacket,
)
from ..aether.events import AetherThreatLevel


# ── AetherNetContextAdapter ───────────────────────────────────────────────────


class AetherNetContextAdapter(IAetherContext):
    """Reports the presence and capability of AetherNet to CircleAI consumers via
    the :class:`IAetherContext` contract.

    Install level is fixed at App for this adapter — AetherNet runs as an
    in-process library; OS-managed instances are surfaced by a separate
    platform-specific adapter.

    :param minimum_required: Minimum AetherNet protocol version the consuming app
        requires. When None, any installed version is considered sufficient.
    :param is_enabled: Whether AetherNet is currently enabled in this process.
        Defaults to True.
    :param protocol_version: The AetherNet wire-protocol version to report as the
        runtime version. Defaults to the runtime's
        ``ProtocolConstants.CurrentProtocolVersion`` (injected here as
        :data:`CURRENT_PROTOCOL_VERSION`), matching the C#
        ``new Version(CurrentProtocolVersion, 0, 0, 0)``.
    """

    def __init__(
        self,
        minimum_required: VersionLike = None,
        is_enabled: bool = True,
        protocol_version: int = CURRENT_PROTOCOL_VERSION,
    ) -> None:
        self._minimum_required = (
            AetherVersion.parse(minimum_required)
            if isinstance(minimum_required, str)
            else minimum_required
        )
        self._is_enabled = is_enabled
        self._runtime_version = AetherVersion(protocol_version, 0, 0, 0)

    @property
    def install_level(self) -> AetherInstallLevel:
        return AetherInstallLevel.APP

    @property
    def is_available(self) -> bool:
        return True

    @property
    def runtime_version(self) -> Optional[AetherVersion]:
        return self._runtime_version

    @property
    def minimum_required(self) -> Optional[AetherVersion]:
        return self._minimum_required

    @property
    def is_sufficient(self) -> bool:
        return self._minimum_required is None or (
            self._runtime_version is not None
            and self._runtime_version >= self._minimum_required
        )

    @property
    def requires_auth(self) -> bool:
        return self.install_level is AetherInstallLevel.OS

    @property
    def is_enabled(self) -> bool:
        return self._is_enabled


# ── AetherNetTelemetryAdapter ─────────────────────────────────────────────────


class AetherNetTelemetryAdapter(IAetherTelemetry):
    """Bridges AetherNet's telemetry bus to CircleAI's :class:`IAetherTelemetry`
    contract. Each subscriber gets an independent AetherNet subscription, so
    disposal cleans up exactly one downstream handle.
    """

    def __init__(self, mesh_telemetry: IAetherNetTelemetry) -> None:
        if mesh_telemetry is None:
            raise ValueError("mesh_telemetry must not be None")
        self._mesh_telemetry = mesh_telemetry

    def subscribe(self, observer: IAetherTelemetryObserver) -> IDisposable:
        if observer is None:
            raise ValueError("observer must not be None")
        bridge = _ObserverBridge(observer)
        return self._mesh_telemetry.subscribe(bridge)


class _ObserverBridge(IAetherNetTelemetryObserver):
    """Receives AetherNet events and forwards them to a CircleAI observer after
    type translation.
    """

    def __init__(self, target: IAetherTelemetryObserver) -> None:
        self._target = target

    def on_node_event(self, e: AetherNetNodeEvent) -> None:
        self._target.on_node_event(EventTranslator.translate_node(e))

    def on_transport_event(self, e: AetherNetTransportEvent) -> None:
        self._target.on_transport_event(EventTranslator.translate_transport(e))

    def on_route_event(self, e: AetherNetRouteEvent) -> None:
        self._target.on_route_event(EventTranslator.translate_route(e))

    def on_security_event(self, e: AetherNetSecurityEvent) -> None:
        self._target.on_security_event(EventTranslator.translate_security(e))

    def on_network_event(self, e: AetherNetNetworkEvent) -> None:
        self._target.on_network_event(EventTranslator.translate_network(e))


# ── AetherNetDirectiveSink (CircleAI -> AetherNet, outbound) ──────────────────


class AetherNetDirectiveSink(ISecurityDirectiveConsumer):
    """Forwards CircleAI security directives to the AetherNet policy engine.
    Implements CircleAI's :class:`ISecurityDirectiveConsumer` so it can be
    registered as a directive sink on the CircleAI side.
    """

    def __init__(self, mesh_consumer: IAetherNetSecurityDirectiveConsumer) -> None:
        if mesh_consumer is None:
            raise ValueError("mesh_consumer must not be None")
        self._mesh_consumer = mesh_consumer

    def on_directive(self, directive: SecurityDirective) -> None:
        if directive is None:
            raise ValueError("directive must not be None")
        mesh_directive = AetherNetSecurityDirective(
            kind=EventTranslator.map_directive_kind_to_mesh(directive.kind),
            target_node_id=directive.target_node_id,
            trust_score_override=directive.trust_score_override,
            threat_level=EventTranslator.map_threat_level_to_mesh(
                directive.threat_level
            ),
            reason=directive.reason,
            duration=directive.duration,
            issued_at=directive.issued_at,
        )
        self._mesh_consumer.on_directive(mesh_directive)


# ── AetherNetInboundDirectiveBridge (AetherNet -> CircleAI, inbound) ──────────


class AetherNetInboundDirectiveBridge(IAetherNetSecurityDirectiveConsumer):
    """Receives AetherNet-side directives and forwards them into CircleAI's
    :class:`ISecurityDirectiveConsumer`. The other half of the bidirectional
    directive pipeline: :class:`AetherNetDirectiveSink` handles outbound, this
    handles inbound.
    """

    def __init__(self, circle_consumer: ISecurityDirectiveConsumer) -> None:
        if circle_consumer is None:
            raise ValueError("circle_consumer must not be None")
        self._circle_consumer = circle_consumer

    def on_directive(self, mesh_directive: AetherNetSecurityDirective) -> None:
        if mesh_directive is None:
            raise ValueError("mesh_directive must not be None")
        circle_directive = SecurityDirective(
            kind=EventTranslator.map_directive_kind_from_mesh(mesh_directive.kind),
            target_node_id=mesh_directive.target_node_id,
            trust_score_override=mesh_directive.trust_score_override,
            threat_level=EventTranslator.map_threat_level(mesh_directive.threat_level),
            reason=mesh_directive.reason,
            duration=mesh_directive.duration,
            issued_at=mesh_directive.issued_at,
        )
        self._circle_consumer.on_directive(circle_directive)


# ── CircleAiAetherNetAiProvider ───────────────────────────────────────────────

_EMPTY_BIASES: Mapping[str, float] = {}
_EMPTY_ROUTES: Sequence[AiRouteSuggestion] = ()


class CircleAiAetherNetAiProvider(IAetherNetAiProvider):
    """Bridges CircleAI's :class:`IAetherIntelligence` to AetherNet's
    :class:`IAetherNetAiProvider` extension seat.

    What CircleAI doesn't yet produce (transport biases, structured route
    suggestions) returns a sensible default that lets the mesh fall back to its
    own logic — no false claim of intelligence we don't have.
    """

    def __init__(self, intelligence: IAetherIntelligence) -> None:
        if intelligence is None:
            raise ValueError("intelligence must not be None")
        self._intelligence = intelligence

    @property
    def is_available(self) -> bool:
        return True

    async def suggest_routes_async(
        self,
        destination_uhid: str,
        payload_bytes: int,
        cancellation_token: object = None,
    ) -> Sequence[AiRouteSuggestion]:
        if not destination_uhid or not destination_uhid.strip():
            return _EMPTY_ROUTES

        advice = await self._intelligence.get_routing_advice_async(
            destination_uhid, cancellation_token
        )
        if advice is None or len(advice.recommended_path) == 0:
            return _EMPTY_ROUTES
        return [AiRouteSuggestion(list(advice.recommended_path), advice.confidence)]

    async def get_transport_biases_async(
        self, payload_bytes: int, cancellation_token: object = None
    ) -> Mapping[str, float]:
        # CircleAI does not yet model per-transport biases. An empty mapping tells
        # AetherNet to use its built-in transport selector without AI adjustment.
        return _EMPTY_BIASES

    async def assess_threat_async(
        self, packet: MeshPacket, cancellation_token: object = None
    ) -> AiThreatLevel:
        if packet is None or not packet.source_uhid or not packet.source_uhid.strip():
            return AiThreatLevel.NONE

        assessment = await self._intelligence.assess_threat_async(
            packet.source_uhid, cancellation_token
        )
        return _map_to_mesh_threat_level(assessment.level)

    async def get_network_health_async(
        self, cancellation_token: object = None
    ) -> AiNetworkHealthReport:
        health = await self._intelligence.get_network_health_async(cancellation_token)
        return AiNetworkHealthReport(
            health.overall_score,
            health.trusted_node_count,
            health.suspicious_node_count,
            health.summary,
            health.generated_at,
        )


def _map_to_mesh_threat_level(level: AetherThreatLevel) -> AiThreatLevel:
    # AetherNet's AiThreatLevel has 4 values (None, Low, Medium, High). CircleAI's
    # AetherThreatLevel has Critical. Fold Critical -> High — the strongest signal
    # the AI seat can carry.
    return {
        AetherThreatLevel.NONE: AiThreatLevel.NONE,
        AetherThreatLevel.LOW: AiThreatLevel.LOW,
        AetherThreatLevel.MEDIUM: AiThreatLevel.MEDIUM,
        AetherThreatLevel.HIGH: AiThreatLevel.HIGH,
        AetherThreatLevel.CRITICAL: AiThreatLevel.HIGH,
    }.get(level, AiThreatLevel.NONE)

File: rust/src/aethernet/adapters.rs
//! aethernet::adapters — the CircleAI ↔ AetherNet adapter family.
//!
//! Rust port of the bridge classes in `CircleAI.AetherNet/*.cs` that wire the
//! CircleAI Aether contracts onto the live mesh runtime boundary
//! ([`crate::aethernet::mesh_extensibility`]):
//!
//!   [`AetherNetContextAdapter`]        — `IAetherContext` over the mesh version.
//!   [`AetherNetTelemetryAdapter`]      — `IAetherTelemetry` over mesh telemetry.
//!   [`AetherNetDirectiveSink`]         — CircleAI → mesh directive forwarder.
//!   [`AetherNetInboundDirectiveBridge`]— mesh → CircleAI directive forwarder.
//!   [`CircleAiAetherNetAiProvider`]    — plugs CircleAI's brain into the mesh
//!                                        AI seat.

use std::collections::BTreeMap;
use std::sync::Arc;

use crate::aether::context::{AetherVersion, IAetherContext};
use crate::aether::events::AetherInstallLevel;
use crate::aether::intelligence::IAetherIntelligence;
use crate::aether::security_layer::{ISecurityDirectiveConsumer, SecurityDirective};
use crate::aether::telemetry::{
    IAetherTelemetry, IAetherTelemetryObserver, TelemetrySubscription,
};

use super::event_translator::{
    map_directive_kind_from_mesh, map_directive_kind_to_mesh, map_threat_level,
    map_threat_level_to_mesh, translate_network_event, translate_node_event, translate_route_event,
    translate_security_event, translate_transport_event,
};
use super::mesh_extensibility::{
    AetherNetNetworkEvent, AetherNetNodeEvent, AetherNetRouteEvent, AetherNetSecurityEvent,
    AetherNetTransportEvent, AiNetworkHealthReport, AiRouteSuggestion, AiThreatLevel,
    IAetherNetAiProvider, IAetherNetTelemetry, IAetherNetTelemetryObserver,
    IMeshSecurityDirectiveConsumer, MeshPacket, MeshSecurityDirective, MeshTelemetrySubscription,
    CURRENT_PROTOCOL_VERSION,
};

// ─────────────────────────────────────────────────────────────────────────────
// AetherNetContextAdapter
// ─────────────────────────────────────────────────────────────────────────────

/// Reports the presence and capability of AetherNet to CircleAI consumers via
/// the [`IAetherContext`] contract. Install level is fixed at
/// [`AetherInstallLevel::App`] — AetherNet runs as an in-process library. Runtime
/// version is `new Version(CURRENT_PROTOCOL_VERSION, 0, 0, 0)`.
#[derive(Debug, Clone)]
pub struct AetherNetContextAdapter {
    minimum_required: Option<AetherVersion>,
    is_enabled: bool,
    runtime_version: AetherVersion,
}

impl AetherNetContextAdapter {
    /// Constructs the adapter. `minimum_required = None` treats any installed
    /// version as sufficient; `is_enabled` defaults to `true` at the call site
    /// (mirrors the C# default parameter).
    pub fn new(minimum_required: Option<AetherVersion>, is_enabled: bool) -> Self {
        Self {
            minimum_required,
            is_enabled,
            runtime_version: AetherVersion::full(CURRENT_PROTOCOL_VERSION, 0, 0, 0),
        }
    }
}

impl Default for AetherNetContextAdapter {
    fn default() -> Self {
        Self::new(None, true)
    }
}

impl IAetherContext for AetherNetContextAdapter {
    fn install_level(&self) -> AetherInstallLevel {
        AetherInstallLevel::App
    }

    fn runtime_version(&self) -> Option<AetherVersion> {
        Some(self.runtime_version)
    }

    fn minimum_required(&self) -> Option<AetherVersion> {
        self.minimum_required
    }

    fn is_enabled(&self) -> bool {
        self.is_enabled
    }

    // C#: `IsAvailable => true` unconditionally (live in-process runtime).
    fn is_available(&self) -> bool {
        true
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// AetherNetTelemetryAdapter
// ─────────────────────────────────────────────────────────────────────────────

/// Bridges AetherNet's telemetry bus to CircleAI's [`IAetherTelemetry`] contract.
/// Each subscriber gets an independent mesh subscription, so disposal cleans up
/// exactly one downstream handle.
pub struct AetherNetTelemetryAdapter {
    mesh_telemetry: Arc<dyn IAetherNetTelemetry>,
}

impl AetherNetTelemetryAdapter {
    pub fn new(mesh_telemetry: Arc<dyn IAetherNetTelemetry>) -> Self {
        Self { mesh_telemetry }
    }
}

impl IAetherTelemetry for AetherNetTelemetryAdapter {
    fn subscribe(&self, observer: Arc<dyn IAetherTelemetryObserver>) -> TelemetrySubscription {
        let bridge: Arc<dyn IAetherNetTelemetryObserver> = Arc::new(ObserverBridge {
            target: observer,
        });
        let mesh_sub = self.mesh_telemetry.subscribe(bridge);
        // Own the mesh subscription inside our handle; dropping the CircleAI
        // handle drops the mesh subscription, unhooking exactly this observer.
        wrap_mesh_subscription(mesh_sub)
    }
}

/// Receives AetherNet events and forwards them to a CircleAI observer after type
/// translation.
struct ObserverBridge {
    target: Arc<dyn IAetherTelemetryObserver>,
}

impl IAetherNetTelemetryObserver for ObserverBridge {
    fn on_node_event(&self, e: &AetherNetNodeEvent) {
        self.target.on_node_event(&translate_node_event(e));
    }
    fn on_transport_event(&self, e: &AetherNetTransportEvent) {
        self.target.on_transport_event(&translate_transport_event(e));
    }
    fn on_route_event(&self, e: &AetherNetRouteEvent) {
        self.target.on_route_event(&translate_route_event(e));
    }
    fn on_security_event(&self, e: &AetherNetSecurityEvent) {
        self.target.on_security_event(&translate_security_event(e));
    }
    fn on_network_event(&self, e: &AetherNetNetworkEvent) {
        self.target.on_network_event(&translate_network_event(e));
    }
}

/// Wraps a mesh subscription in a CircleAI [`TelemetrySubscription`] so the outer
/// handle's lifetime governs the inner one.
fn wrap_mesh_subscription(mesh_sub: MeshTelemetrySubscription) -> TelemetrySubscription {
    use std::sync::Mutex;
    // Move the mesh sub into a closure captured by the CircleAI handle's remover.
    let holder = Arc::new(Mutex::new(Some(mesh_sub)));
    TelemetrySubscription::from_remover(move || {
        // Dropping the held mesh subscription unhooks the mesh observer.
        holder.lock().unwrap().take();
    })
}

// ─────────────────────────────────────────────────────────────────────────────
// AetherNetDirectiveSink   (CircleAI → mesh, outbound)
// ─────────────────────────────────────────────────────────────────────────────

/// Forwards CircleAI security directives to the AetherNet policy engine.
/// Implements CircleAI's [`ISecurityDirectiveConsumer`] so it can be registered
/// as a directive sink on the CircleAI side.
pub struct AetherNetDirectiveSink {
    mesh_consumer: Arc<dyn IMeshSecurityDirectiveConsumer>,
}

impl AetherNetDirectiveSink {
    pub fn new(mesh_consumer: Arc<dyn IMeshSecurityDirectiveConsumer>) -> Self {
        Self { mesh_consumer }
    }
}

impl ISecurityDirectiveConsumer for AetherNetDirectiveSink {
    fn on_directive(&self, directive: &SecurityDirective) {
        let mesh_directive = MeshSecurityDirective {
            kind: map_directive_kind_to_mesh(directive.kind),
            target_node_id: directive.target_node_id.clone(),
            trust_score_override: directive.trust_score_override,
            threat_level: map_threat_level_to_mesh(directive.threat_level),
            reason: directive.reason.clone(),
            duration: directive.duration,
            issued_at: directive.issued_at,
        };
        self.mesh_consumer.on_directive(&mesh_directive);
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// AetherNetInboundDirectiveBridge   (mesh → CircleAI, inbound)
// ─────────────────────────────────────────────────────────────────────────────

/// Receives mesh-side [`MeshSecurityDirective`]s and forwards them into CircleAI's
/// [`ISecurityDirectiveConsumer`] (typically a `MeshDirectiveStore`). The inverse
/// of [`AetherNetDirectiveSink`].
pub struct AetherNetInboundDirectiveBridge {
    circle_consumer: Arc<dyn ISecurityDirectiveConsumer>,
}

impl AetherNetInboundDirectiveBridge {
    pub fn new(circle_consumer: Arc<dyn ISecurityDirectiveConsumer>) -> Self {
        Self { circle_consumer }
    }
}

impl IMeshSecurityDirectiveConsumer for AetherNetInboundDirectiveBridge {
    fn on_directive(&self, mesh_directive: &MeshSecurityDirective) {
        let circle_directive = SecurityDirective::new(
            map_directive_kind_from_mesh(mesh_directive.kind),
            mesh_directive.target_node_id.clone(),
            mesh_directive.trust_score_override,
            map_threat_level(mesh_directive.threat_level),
            mesh_directive.reason.clone(),
            mesh_directive.duration,
            mesh_directive.issued_at,
        );
        self.circle_consumer.on_directive(&circle_directive);
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// CircleAiAetherNetAiProvider
// ─────────────────────────────────────────────────────────────────────────────

/// Bridges CircleAI's [`IAetherIntelligence`] to AetherNet's AI-provider seat
/// ([`IAetherNetAiProvider`]). The mesh routing layer asks for route advice,
/// threat assessments, and network health; each call forwards to CircleAI's
/// intelligence surface and translates the result. Signals CircleAI does not yet
/// produce (transport biases, structured route suggestions beyond the path)
/// return sensible defaults so the mesh falls back to its own logic.
pub struct CircleAiAetherNetAiProvider {
    intelligence: Arc<dyn IAetherIntelligence>,
}

impl CircleAiAetherNetAiProvider {
    pub fn new(intelligence: Arc<dyn IAetherIntelligence>) -> Self {
        Self { intelligence }
    }

    /// AetherNet's `AiThreatLevel` has only 4 values (None..High). CircleAI's
    /// `AetherThreatLevel` has Critical; fold Critical → High because that's the
    /// strongest signal the AI seat can carry.
    fn map_to_mesh_threat_level(
        l: crate::aether::events::AetherThreatLevel,
    ) -> AiThreatLevel {
        use crate::aether::events::AetherThreatLevel as L;
        match l {
            L::None => AiThreatLevel::None,
            L::Low => AiThreatLevel::Low,
            L::Medium => AiThreatLevel::Medium,
            L::High => AiThreatLevel::High,
            L::Critical => AiThreatLevel::High,
        }
    }
}

impl IAetherNetAiProvider for CircleAiAetherNetAiProvider {
    fn is_available(&self) -> bool {
        true
    }

    fn suggest_routes(
        &self,
        destination_uhid: &str,
        _payload_bytes: i32,
    ) -> Vec<AiRouteSuggestion> {
        if destination_uhid.trim().is_empty() {
            return Vec::new();
        }
        let advice = self.intelligence.get_routing_advice(destination_uhid);
        if advice.recommended_path.is_empty() {
            return Vec::new();
        }
        vec![AiRouteSuggestion {
            path: advice.recommended_path,
            confidence: advice.confidence,
        }]
    }

    fn get_transport_biases(&self, _payload_bytes: i32) -> BTreeMap<String, f64> {
        // CircleAI does not yet model per-transport biases — empty tells
        // AetherNet to use its built-in selector without AI adjustment.
        BTreeMap::new()
    }

    fn assess_threat(&self, packet: &MeshPacket) -> AiThreatLevel {
        if packet.source_uhid.trim().is_empty() {
            return AiThreatLevel::None;
        }
        let assessment = self.intelligence.assess_threat(&packet.source_uhid);
        Self::map_to_mesh_threat_level(assessment.level)
    }

    fn get_network_health(&self) -> AiNetworkHealthReport {
        let health = self.intelligence.get_network_health();
        AiNetworkHealthReport {
            overall_score: health.overall_score,
            trusted_node_count: health.trusted_node_count,
            suspicious_node_count: health.suspicious_node_count,
            summary: health.summary,
            generated_at: health.generated_at,
        }
    }
}
